### PR TITLE
#53 create get product/list endpoint

### PR DESCRIPTION
**Pull request by vita792**

### CODE REVIEW

File: src/main/java/com/megastore/controller/ProductController.java [hunk 1]
```diff
@@ -29,12 +31,31 @@ public ProductController(ProductFacade productFacade) {
...
+        return ResponseEntity.ok(new ArrayList<>(productFacade.findHotProducts(limit, catId)));
+    }
 
+    @GetMapping("/list")
+    public ResponseEntity<Collection<ProductPLPDto>> getAllProducts(
+            @RequestParam(defaultValue = "9") int limit,
+            @RequestParam(defaultValue = "1") long subcategoryId,
+            @RequestParam(defaultValue = "UPDATED") Sorting sort,
+            @RequestParam(defaultValue = "0.00") Double priceFrom,
+            @RequestParam(defaultValue = "1000000.00") Double priceTo,
```
Comment (dimalian) [inline src/main/java/com/megastore/controller/ProductController.java]: you shouldn't have a default value for price. If there is no price, you need to return all items, even if the price is over 100000.

File: src/main/java/com/megastore/controller/ProductController.java [hunk 2]
```diff
@@ -29,12 +31,31 @@ public ProductController(ProductFacade productFacade) {
...
+    ) {
+        Collection<ProductPLPDto> products;
+        if ((sort.toString()).equals("ALPHABETIC")) {
+            products = new ArrayList<>(productFacade.findAllSortedByNameASC(limit, subcategoryId, priceFrom, priceTo, brand));
+        } else if ((sort.toString()).equals("REVERSED")) {
+            products = new ArrayList<>(productFacade.findAllSortedByNameDESC(limit, subcategoryId, priceFrom, priceTo, brand));
+        } else if ((sort.toString()).equals("CHEAP")) {
+            products = new ArrayList<>(productFacade.findAllSortedByPriceASC(limit, subcategoryId, priceFrom, priceTo, brand));
+        } else if ((sort.toString()).equals("EXPENSIVE")) {
+            products = new ArrayList<>(productFacade.findAllSortedByPriceDESC(limit, subcategoryId, priceFrom, priceTo, brand));
```
Comment (dimalian) [inline src/main/java/com/megastore/controller/ProductController.java]: As I mentioned in the meeting, to sort a product you need to get 2 parameters:
**sortBy**: "price" or "alphabetical"
**orderBy**: "ASC" or "DESC"
The first parameter indicates what type to sort by.
The second parameter tells you in which direction you will sort.

File: src/main/java/com/megastore/facade/ProductFacade.java [hunk 3]
```diff
@@ -10,8 +10,18 @@
...
-    Collection<ProductPLPDto> findAll();
+    Collection<ProductPLPDto> findAll(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand);
+
+    Collection<ProductPLPDto> findAllSortedByNameDESC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand);
+
+    Collection<ProductPLPDto> findAllSortedByNameASC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand);
+
+    Collection<ProductPLPDto> findAllSortedByPriceDESC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand);
+
+    Collection<ProductPLPDto> findAllSortedByPriceASC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand);
```
Comment (dimalian) [inline src/main/java/com/megastore/facade/ProductFacade.java]: There shouldn't be 5 different functions for each sort type. Why not do the same as with filtering parameters?

File: src/main/java/com/megastore/facade/impl/ProductFacadeImpl.java [hunk 4]
```diff
@@ -26,8 +27,28 @@ public ProductPDPDto findById(Long id) {
...
+    }
+
+    @Override
+    public Collection<ProductPLPDto> findAllSortedByPriceDESC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand) {
+        return productService.findAllSortedByPriceDESC(limit, subcategoryId, priceFrom, priceTo, brand).stream().map(ProductPLPDto::new).toList();
+    }
+
+    @Override
+    public Collection<ProductPLPDto> findAllSortedByPriceASC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand) {
+        return productService.findAllSortedByPriceASC(limit, subcategoryId, priceFrom, priceTo, brand).stream().map(ProductPLPDto::new).toList();
```
Comment (dimalian) [inline src/main/java/com/megastore/facade/impl/ProductFacadeImpl.java]: How many functions would you create if we had 10 parameters for sorting?

File: src/main/java/com/megastore/repository/ProductRepository.java [hunk 5]
```diff
@@ -14,4 +15,78 @@ public interface ProductRepository extends BaseRepository<Product> {
...
+            "JOIN categories c ON s.category_id = c.id " +
+            "JOIN images img ON img.product_id = p.id " +
+            "WHERE s.id = :subcategoryId " +
+            "AND p.product_price BETWEEN :priceFrom AND :priceTo " +
+            "AND p.product_brand = :brand " +
+            "ORDER BY p.product_price ASC, p.updated DESC " +
+            "LIMIT :limit ",
+            nativeQuery = true)
+    Collection<Product> findAllSortedByPriceASC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand);
+
```
Comment (dimalian) [inline src/main/java/com/megastore/repository/ProductRepository.java]: You should feel like something went wrong when you write almost identical code 6 times.  Please, find better solution.

File: src/main/java/com/megastore/service/impl/ProductServiceImpl.java [hunk 6]
```diff
@@ -36,7 +36,31 @@ public Optional<Product> findById(Long id) {
...
+    @Override
+    public Collection<Product> findAllSortedByPriceDESC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand) {
+        return productRepository.findAllSortedByPriceDESC(limit, subcategoryId, priceFrom, priceTo, brand);
+    }
+
+    @Transactional(readOnly = true)
+    @Override
+    public Collection<Product> findAllSortedByPriceASC(int limit, long subcategoryId, Double priceFrom, Double priceTo, String brand) {
+        return productRepository.findAllSortedByPriceASC(limit, subcategoryId, priceFrom, priceTo, brand);
     }
```
Comment (dimalian) [inline src/main/java/com/megastore/service/impl/ProductServiceImpl.java]: Same here

**Review by dimalian — CHANGES_REQUESTED** [review]

**Review by dimalian — APPROVED** [review]

**Review by TarasFedorenko — APPROVED** [review]

**Pull request merged**